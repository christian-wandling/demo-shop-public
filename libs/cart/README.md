# `@demo-shop/cart`

This library handles all aspects of the shopping cart feature including product selection, cart management, and the checkout process.

## Features

- **Shopping Cart Management**

  - Add/remove items
  - Update quantities

- **Checkout Process**

  - Address management
  - Order review
  - Order confirmation

- **State Management**
  - Signal based state
  - Synchronization with backend

## API Reference

For a complete API reference, see the [Compodoc documentation](https://christian-wandling.github.io/demo-shop-public/compodoc/frontend).

## Running Unit Tests

Run `nx test cart` to execute the unit tests.
