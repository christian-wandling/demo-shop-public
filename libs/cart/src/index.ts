export * from './lib/cart.routes';
export * from './lib/cart.facade';
export * from './lib/components/cart/cart.component';
export * from './lib/components/checkout/checkout.component';
export * from './lib/components/cart-icon/cart-icon.component';
