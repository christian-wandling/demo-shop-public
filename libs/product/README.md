# `@demo-shop/product`

A library for product management. It enables displaying and filtering the product catalog and add them to the shopping cart.

## Features

- **Product Catalog**

  - Response grid view of product items

- **Product Details**

  - Detailed product information

- **Shopping Integration**

  - One-click add to cart functionality

- **Product Filtering**

  - Realtime full text search filter
  - Supports text and category filtering

- **State Management**
  - Signal-based reactive store
  - Synchronization with backend

## API Reference

For a complete API reference, see the [Compodoc documentation](https://christian-wandling.github.io/demo-shop-public/compodoc/frontend).

## Running unit tests

Run `nx test product` to execute the unit tests.
