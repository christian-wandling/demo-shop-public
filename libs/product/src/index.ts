export * from './lib/product.routes';
export * from './lib/models/product-filter';
export * from './lib/product.facade';
export * from './lib/components/product-search/product-search.component';
