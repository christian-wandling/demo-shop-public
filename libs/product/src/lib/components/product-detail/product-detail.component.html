@if (product()) {
<div class="product-detail pt-6">
  <!-- Image gallery -->
  <div class="mx-auto mt-6 max-w-2xl sm:px-6 lg:grid lg:max-w-7xl lg:grid-cols-3 lg:gap-x-8 lg:px-8">
    <div class="aspect-h-1 aspect-w-1 overflow-hidden sm:rounded-lg lg:block">
      <img
        [ngSrc]="product().thumbnail.uri ?? ''"
        [alt]="product().name"
        onerror="this.src='images/placeholder-image.jpg'"
        class="h-full w-full object-cover object-center group-hover:opacity-75"
        fill="" />
    </div>
  </div>

  <!-- Product info -->
  <div
    class="mx-auto max-w-2xl px-4 pb-16 pt-10 sm:px-6 lg:grid lg:max-w-7xl lg:grid-cols-3 lg:grid-rows-[auto,auto,1fr] lg:gap-x-8 lg:px-8 lg:pb-24 lg:pt-16">
    <div class="lg:col-span-2 lg:border-r lg:border-gray-200 lg:pr-8">
      <h1 class="text-2xl font-bold tracking-tight text-gray-900 sm:text-3xl">{{ product().name }}</h1>
    </div>

    <!-- Options -->
    <div class="mt-4 lg:row-span-3 lg:mt-0">
      <h2 class="sr-only">Product information</h2>
      <p class="text-3xl tracking-tight text-gray-900">
        {{ product().price | currency }}
      </p>

      <button
        (click)="addToCart(product().id)"
        [disabled]="!addButtonEnabled()"
        class="btn-add mt-10 flex w-full items-center justify-center rounded-md border border-transparent px-8 py-3 text-base font-medium"
        [ngClass]="{
          'bg-gray-100 text-gray-800 pointer-events-none': !addButtonEnabled(),
          'bg-indigo-600 text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2':
            addButtonEnabled()
        }">
        Add to cart
      </button>
    </div>

    <div class="py-10 lg:col-span-2 lg:col-start-1 lg:border-r lg:border-gray-200 lg:pb-16 lg:pr-8 lg:pt-6">
      <div class="product-description text-sm text-gray-600">{{ product().description }}</div>
    </div>
  </div>
</div>
}
