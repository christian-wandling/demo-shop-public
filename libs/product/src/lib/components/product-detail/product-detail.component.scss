.product-description ::ng-deep {
  p, h3 {
    margin: 1rem 0;
  }

  p:first-child {
    font-size: 1rem;
    line-height: 1.5rem;
  }

  h3 {
    color: #111827;
    font-weight: 500;
    font-size: 0.875rem;
    line-height: 1.25rem;
    margin-top: 2.5rem;
  }

  ul {
    list-style-type: disc;
    list-style-position: outside;
    padding-left: 1rem;
  }

  li {
    &:not(:last-child) {
      margin-bottom: 0.25rem;
    }

    &::marker {
      color: #9ca3af;
    }
  }
}
