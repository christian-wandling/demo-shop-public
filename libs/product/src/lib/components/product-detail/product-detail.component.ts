import { ChangeDetectionStrategy, Component, inject } from '@angular/core';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { ProductFacade } from '../../product.facade';
import { ActivatedRoute } from '@angular/router';
import { CartFacade } from '@demo-shop/cart';

@Component({
  selector: 'lib-product-detail',
  standalone: true,
  imports: [CommonModule, NgOptimizedImage],
  templateUrl: './product-detail.component.html',
  styleUrl: './product-detail.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
  host: {
    class: 'bg-white',
  },
})
export class ProductDetailComponent {
  readonly #activatedRoute = inject(ActivatedRoute);
  readonly #productFacade = inject(ProductFacade);
  readonly product = this.#productFacade.getById(this.#activatedRoute.snapshot.params['id']);
  readonly #cartFacade = inject(CartFacade);
  readonly addButtonEnabled = this.#cartFacade.getHasShoppingSession();

  addToCart(id: string): void {
    this.#cartFacade.addItem(id);
  }
}
