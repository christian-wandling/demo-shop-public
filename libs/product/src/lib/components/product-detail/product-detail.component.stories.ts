import { applicationConfig, Meta, StoryObj } from '@storybook/angular';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { ActivatedRoute } from '@angular/router';
import { importProvidersFrom, signal } from '@angular/core';
import { ProductDetailComponent } from './product-detail.component';
import { ProductFacade } from '../../product.facade';
import { ImageResponse, ProductResponse } from '@demo-shop/api';
import { CartFacade } from '@demo-shop/cart';
import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';
import { faker } from '@faker-js/faker';
import { provideImageLoader } from '@demo-shop/shared';

const name = faker.commerce.productName();
const image: ImageResponse = {
  name,
  uri: faker.image.urlPicsumPhotos(),
};

const product: ProductResponse = {
  id: 1,
  name,
  price: faker.number.int({ min: 1, max: 20 }) * 10 - 1,
  description: faker.commerce.productDescription(),
  categories: [faker.commerce.department()],
  images: [image],
  thumbnail: image,
};

const mockProductFacade = (product: ProductResponse) => ({
  getById: () => signal(product),
  fetchById: () => {
    return;
  },
});

const mockCartFacade = (hasShoppingSession: boolean) => ({
  getHasShoppingSession: () => signal(hasShoppingSession),
  addItem: () => {
    return;
  },
});

const mockActivatedRoute = {
  snapshot: {
    params: {
      id: 1,
    },
  },
};

const meta: Meta<ProductDetailComponent> = {
  component: ProductDetailComponent,
  title: 'Product/ProductDetailComponent',
  decorators: [
    applicationConfig({
      providers: [
        importProvidersFrom(CommonModule, NgOptimizedImage),
        provideImageLoader(),
        { provide: ProductFacade, useValue: mockProductFacade(product) },
        { provide: ActivatedRoute, useValue: mockActivatedRoute },
        { provide: CartFacade, useValue: mockCartFacade(false) },
      ],
    }),
  ],
  argTypes: {
    addToCart: { action: 'addToCart', control: false },
  },
  parameters: {
    controls: {
      exclude: ['#activatedRoute', '#productFacade', '#cartFacade', 'product', 'addButtonEnabled', 'ngOnInit'],
    },
  },
};
export default meta;
type Story = StoryObj<ProductDetailComponent>;

export const Unauthenticated: Story = {
  play: async ({ canvasElement, args }) => {
    const button = within(canvasElement).getByRole('button');
    expect(button).toBeDisabled();
  },
};

export const Authenticated: Story = {
  decorators: [
    applicationConfig({
      providers: [{ provide: CartFacade, useValue: mockCartFacade(true) }],
    }),
  ],
  play: async ({ canvasElement, args }) => {
    const button = within(canvasElement).getByRole('button');
    expect(button).not.toBeDisabled();
  },
};
