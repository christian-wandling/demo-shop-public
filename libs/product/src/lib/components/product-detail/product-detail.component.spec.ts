import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ProductDetailComponent } from './product-detail.component';
import { ActivatedRoute } from '@angular/router';
import { ProductFacade } from '../../product.facade';
import { CartFacade } from '@demo-shop/cart';
import { signal } from '@angular/core';
import { By } from '@angular/platform-browser';

describe('ProductDetailComponent', () => {
  const hasShoppingSession = signal(false);

  let component: ProductDetailComponent;
  let fixture: ComponentFixture<ProductDetailComponent>;
  let cartFacade: CartFacade;

  const mockProduct = signal({
    id: 1,
    name: 'Product 1',
    price: 10,
    description: '',
    categories: [],
    images: [
      {
        name: 'name',
        uri: 'uri',
      },
    ],
  });

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [ProductDetailComponent],
      providers: [
        {
          provide: ActivatedRoute,
          useValue: {
            snapshot: {
              params: {
                id: 1,
              },
            },
          },
        },
        {
          provide: ProductFacade,
          useValue: {
            getById: jest.fn().mockReturnValue(mockProduct),
            fetchById: jest.fn(),
          },
        },
        {
          provide: CartFacade,
          useValue: {
            getHasShoppingSession: jest.fn().mockReturnValue(hasShoppingSession),
            addItem: jest.fn(),
          },
        },
      ],
    }).compileComponents();

    cartFacade = TestBed.inject(CartFacade);
    fixture = TestBed.createComponent(ProductDetailComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should render the product details', () => {
    expect(fixture.debugElement.query(By.css('.product-detail'))).toMatchSnapshot();
  });

  describe('addButton', () => {
    it('should be enabled if a shoppingSession exists', () => {
      hasShoppingSession.set(true);
      fixture.detectChanges();

      const btn = fixture.debugElement.query(By.css('.btn-add'));

      expect(btn.nativeElement.disabled).toBe(false);
    });

    it('should be disabled if no shoppingSession exists', () => {
      hasShoppingSession.set(false);
      fixture.detectChanges();

      const btn = fixture.debugElement.query(By.css('.btn-add'));

      expect(btn.nativeElement.disabled).toBe(true);
    });
  });

  it('should add the item to the cart', () => {
    component.addToCart(1);

    expect(cartFacade.addItem).toHaveBeenCalledWith(1);
  });
});
