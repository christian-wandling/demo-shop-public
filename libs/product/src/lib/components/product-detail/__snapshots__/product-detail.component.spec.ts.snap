// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetailComponent should match the snapshot 1`] = `
<lib-product-detail
  addButtonEnabled={[Function Function]}
  product={[Function Function]}
>
  <div
    class="product-detail pt-6"
  >
    <div
      class="mx-auto mt-6 max-w-2xl sm:px-6 lg:grid lg:max-w-7xl lg:grid-cols-3 lg:gap-x-8 lg:px-8"
    >
      <div
        class="aspect-h-1 aspect-w-1 hidden overflow-hidden rounded-lg lg:block"
      >
        <img
          alt="Product 1"
          class="h-full w-full object-cover object-center group-hover:opacity-75"
          fetchpriority="auto"
          fill=""
          loading="lazy"
          ng-img="true"
          sizes="100vw"
          src="uri"
          style="position: absolute; width: 100%; height: 100%; inset: 0;"
        />
      </div>
    </div>
    <div
      class="mx-auto max-w-2xl px-4 pb-16 pt-10 sm:px-6 lg:grid lg:max-w-7xl lg:grid-cols-3 lg:grid-rows-[auto,auto,1fr] lg:gap-x-8 lg:px-8 lg:pb-24 lg:pt-16"
    >
      <div
        class="lg:col-span-2 lg:border-r lg:border-gray-200 lg:pr-8"
      >
        <h1
          class="text-2xl font-bold tracking-tight text-gray-900 sm:text-3xl"
        >
          Product 1
        </h1>
      </div>
      <div
        class="mt-4 lg:row-span-3 lg:mt-0"
      >
        <h2
          class="sr-only"
        >
          Product information
        </h2>
        <p
          class="text-3xl tracking-tight text-gray-900"
        >
           $10.00 
        </p>
        <button
          class="btn-add mt-10 flex w-full items-center justify-center rounded-md border border-transparent px-8 py-3 text-base font-medium bg-gray-100 text-gray-800 pointer-events-none"
          disabled=""
        >
           Add to cart 
        </button>
      </div>
      <div
        class="py-10 lg:col-span-2 lg:col-start-1 lg:border-r lg:border-gray-200 lg:pb-16 lg:pr-8 lg:pt-6"
      >
        <div
          class="product-description text-sm text-gray-600"
        >
        </div>
      </div>
    </div>
  </div>
</lib-product-detail>
`;

exports[`ProductDetailComponent should render the product details 1`] = `
DebugElement {
  "nativeNode": <div
    class="product-detail pt-6"
  >
    <div
      class="mx-auto mt-6 max-w-2xl sm:px-6 lg:grid lg:max-w-7xl lg:grid-cols-3 lg:gap-x-8 lg:px-8"
    >
      <div
        class="aspect-h-1 aspect-w-1 hidden overflow-hidden rounded-lg lg:block"
      >
        <img
          alt="Product 1"
          class="h-full w-full object-cover object-center group-hover:opacity-75"
          fetchpriority="auto"
          fill=""
          loading="lazy"
          ng-img="true"
          sizes="100vw"
          src="uri"
          style="position: absolute; width: 100%; height: 100%; inset: 0;"
        />
      </div>
    </div>
    <div
      class="mx-auto max-w-2xl px-4 pb-16 pt-10 sm:px-6 lg:grid lg:max-w-7xl lg:grid-cols-3 lg:grid-rows-[auto,auto,1fr] lg:gap-x-8 lg:px-8 lg:pb-24 lg:pt-16"
    >
      <div
        class="lg:col-span-2 lg:border-r lg:border-gray-200 lg:pr-8"
      >
        <h1
          class="text-2xl font-bold tracking-tight text-gray-900 sm:text-3xl"
        >
          Product 1
        </h1>
      </div>
      <div
        class="mt-4 lg:row-span-3 lg:mt-0"
      >
        <h2
          class="sr-only"
        >
          Product information
        </h2>
        <p
          class="text-3xl tracking-tight text-gray-900"
        >
           $10.00 
        </p>
        <button
          class="btn-add mt-10 flex w-full items-center justify-center rounded-md border border-transparent px-8 py-3 text-base font-medium bg-gray-100 text-gray-800 pointer-events-none"
          disabled=""
        >
           Add to cart 
        </button>
      </div>
      <div
        class="py-10 lg:col-span-2 lg:col-start-1 lg:border-r lg:border-gray-200 lg:pb-16 lg:pr-8 lg:pt-6"
      >
        <div
          class="product-description text-sm text-gray-600"
        >
          
        </div>
      </div>
    </div>
  </div>,
}
`;
